### PR TITLE
feat(ci): cache Docker deps layer

### PR DESCRIPTION
**Pull request by clemsos**
# Description

To speed up ci, we cache a Docker image with all node packages installed after each master branch run - and pull from the cached image when running tests on PRs.

This requires #10241 to be merged first - as we are targeting the `deps` stage of the Dockerfile that contains all pre-installed node packages 

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by julien51** [comment]
Hey @clemsos do you think this one is ready?

**Comment by clemsos** [comment]
> Hey @clemsos do you think this one is ready?

Not far from ready but needs testing. Needed the new docker file to work first

File: docker/docker-compose.ci.yml [hunk 1]
```diff
@@ -13,13 +13,23 @@ services:
     build:
       context: ../
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:17]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 2]
```diff
@@ -13,13 +13,23 @@ services:
...
   
   subgraph: # Subgraph unit testing
     image: subgraph
     platform: linux/amd64
     build:
       context: ../
+      target: subgraph
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:29]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 3]
```diff
@@ -28,6 +38,10 @@ services:
     build:
       context: ../
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:42]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 4]
```diff
@@ -40,6 +54,10 @@ services:
...
```
Comment (julien51) [inline docker/docker-compose.ci.yml:58]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 5]
```diff
@@ -50,6 +68,11 @@ services:
     build:
       context: ../
       target: dev
+
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:73]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 6]
```diff
@@ -60,6 +83,10 @@ services:
     build:
       context: ../
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:87]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 7]
```diff
@@ -68,6 +95,10 @@ services:
...
```
Comment (julien51) [inline docker/docker-compose.ci.yml:99]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 8]
```diff
@@ -76,13 +107,22 @@ services:
...
```
Comment (julien51) [inline docker/docker-compose.ci.yml:111]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 9]
```diff
@@ -76,13 +107,22 @@ services:
...
     env_file:
       - ./${UNLOCK_ENV}.env
+
   contracts: # Container for @unlock-protocol/contracts
     image: contracts
     build:
       context: ../
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:123]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 10]
```diff
@@ -91,6 +131,10 @@ services:
...
```
Comment (julien51) [inline docker/docker-compose.ci.yml:135]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 11]
```diff
@@ -99,7 +143,10 @@ services:
     build:
       context: ../
       target: dev
-
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:147]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 12]
```diff
@@ -129,6 +180,10 @@ services:
     build:
       context: ../
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.ci.yml:184]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.integration.yml [hunk 13]
```diff
@@ -33,6 +33,10 @@ services:
     build:
       context: ../
       target: dev
+      cache_from: 
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.integration.yml:37]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.integration.yml [hunk 14]
```diff
@@ -55,6 +59,10 @@ services:
...
```
Comment (julien51) [inline docker/docker-compose.integration.yml:63]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.override.yml [hunk 15]
```diff
@@ -24,10 +24,11 @@ services:
...
     build:
       context: ../
-      target: dev
+      target: dev  
+      args:
+        BUILD_DIR: locksmith
       cache_from:
-        - locksmith
-        - unlockprotocol/locksmith:master
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.override.yml:31]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.override.yml [hunk 16]
```diff
@@ -55,8 +56,7 @@ services:
       context: ../
       target: dev
       cache_from:
-        - locksmith
-        - unlockprotocol/locksmith:master
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.override.yml:59]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.override.yml [hunk 17]
```diff
@@ -74,11 +74,16 @@ services:
...
       - postgres
       - eth-node
+
   # deploy the unlock graph on graph-node, then stop after execution
   subgraph:
     build:
       context: ../
       target: dev
+      cache_from:
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.override.yml:84]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.override.yml [hunk 18]
```diff
@@ -100,6 +105,8 @@ services:
     image: wedlocks
     build:
       target: dev
+      cache_from:
+        - unlocktravis/unlock-dev
```
Comment (julien51) [inline docker/docker-compose.override.yml:109]:
```suggestion
        - unlockprotocol/unlock-dev
```

File: docker/docker-compose.ci.yml [hunk 19]
```diff
@@ -13,13 +13,22 @@ services:
...
+        - unlockprotocol/unlock-dev
+      args:
+        BUILD_DIR: smart-contracts
   
   subgraph: # Subgraph unit testing
     image: subgraph
     platform: linux/amd64
     build:
       context: ../
+      target: subgraph
```
Comment (julien51) [inline docker/docker-compose.ci.yml:26]: ?

Comment (julien51) [reply docker/docker-compose.ci.yml:26]: Should we remove the `target: dev` from below

Comment (clemsos) [reply docker/docker-compose.ci.yml]: actually the subgraph target was removed

**Review by julien51 — APPROVED** [review]

**Comment by julien51** [comment]
Image is now on Docker hub! https://hub.docker.com/repository/docker/unlockprotocol/unlock-dev

File: .github/workflows/master.yml [hunk 20]
```diff
@@ -4,6 +4,7 @@ on:
   push:
     branches:
       - master
+      - ci-docker-cache
```
Comment (julien51) [inline .github/workflows/master.yml:7]:
```suggestion
```

Comment (julien51) [reply .github/workflows/master.yml:7]: Let's cleanup before merging!

**Comment by julien51** [comment]
Ok this passed https://github.com/unlock-protocol/unlock/actions/runs/3584292930/jobs/6030775887

Now running to see if it pulls!

**Comment by julien51** [comment]
We need to add `yarn` command to make sure packages are up to date, just in case a branch might have added/removed them.
If no change i dependencies it should be instantaneous

File: .github/workflows/master.yml (hunk 20 above)
Comment (julien51) [inline .github/workflows/master.yml]:
```suggestion
      - ci-docker-cache
```

**Pull request merged**